lib/sycamore/sycamore/connectors/neo4j/neo4j_writer.py

```diff
@@ -0,0 +1,343 @@
+            os.makedirs(os.path.dirname(csv_path), exist_ok=True)
+            with open(csv_path, "w", newline="") as file:
+                writer = csv.writer(file)
+                writer.writerow(headers)
+
+
+class Neo4jWriteCSV(MapBatch, Write):
+    def __init__(self, plan: Node, client_params: Neo4jWriterClientParams, **kwargs):
+        self._import_dir = client_params.import_dir
+        self._name = "Neo4jWriter"
+        super().__init__(plan, f=self._write_docs_tt, **kwargs)
+
+    def _parse_docs(self, docs):
+        include = ["type", "bbox", "text_representation"]
+        nodes = defaultdict(list)
+        relationships = defaultdict(lambda: defaultdict(list))
+        for doc in docs:
+            if "label" not in doc.data or "doc_id" not in doc.data:
+                continue
+            node = {
+                "uuid:ID": doc.doc_id,
+            }
+            if doc["type"] == "table":
+                doc["text_representation"] = doc["table"].to_csv()
+            for key, value in doc.data.items():
+                # add flatten properties to node
+                if key == "properties":
+                    if isinstance(value, dict) and value:
+                        properties_flat = pd.json_normalize(value, sep="_").to_dict(orient="records")[0]
+                        for property_key, property_value in properties_flat.items():
+                            if property_key not in include:
+                                node[property_key] = property_value
+                # add included fields to node
+                if key in include:
+                    node[key] = value
+            nodes[doc.data["label"]].append(node)
+
+            for key, value in doc.data["relationships"].items():
+                rel = {
+                    "uuid:ID": key,
+                    ":START_ID": value["START_ID"],
+                    ":END_ID": value["END_ID"],
+                    ":TYPE": value["TYPE"],
+                }
+                for key, value in value["properties"].items():
+                    if key not in ["uuid:ID", ":START_ID", ":END_ID", ":TYPE"]:
+                        rel[key] = value
+
+                relationships[value["START_LABEL"]][value["END_LABEL"]].append(rel)
+        return nodes, relationships
+
+    def _write_nodes_csv(self, nodes):
+        for node_label, rels in nodes.items():
+            csv_path = os.path.expanduser(f"{self._import_dir}/sycamore/nodes/{node_label}.csv")
+            headers = None
+            with open(csv_path, mode="r", newline="") as file:
+                csv_reader = csv.reader(file)
+                headers = sorted(set(next(csv_reader)))
+
+            with open(csv_path, "a", newline="") as file:
+                writer = csv.writer(file)
+                for entry in rels:
+                    row = [entry.get(header, "") for header in headers]
+                    writer.writerow(row)
```

Comment:
q: is there a race condition here that could give you weird spliced lines?

Reply:
yep, looks like there could be a race condition. Have not run into one thus far but should protect against it. Will use filesystem lock ```fcntl.lockf``` to deal with this.